Dynamic TFLOPS Calculation (In response to #243)

snico432 opened the pull request:
I attempted to implement a dynamic TFLOPS calculation as a fallback in case the device is not found in the lookup table. I know that PyTorch is not yet a dependency for the project but I saw some discussion in #139 that it will be soon. Please forgive me if this PR is incorrectly formatted, it's my first time attempting to contribute to an open source project, but I was motivated to because I find this project really cool! Also, please let me know if there are any changes I should make to the code.

AlexCheema commented:
Yes!! I was hoping someone would do this. Great work! I haven't taken a proper look / tested yet, but I love to idea of dynamically calculating FLOPS based on a quick benchmark.

I added a $200 retrospective bounty for this which will be paid once merged

AlexCheema commented:
One suggestion: I don't want to force torch as a dependency. If we could lean on the existing `InferenceEngine` infrastructure we have, that would be great. Perhaps each `InferenceEngine` can implement a `benchmark` function?

AlexCheema commented:
<img width="223" alt="Screenshot 2024-10-07 at 01 05 01" src="https://github.com/user-attachments/assets/a2dc9e6a-c1cd-4504-821b-f2c35a5f62f1">

Getting this output now after 

```
exo --inference-engine pytorch --run-model llama-3.1-8b
```

snico432 commented:
> <img alt="Screenshot 2024-10-07 at 01 05 01" width="223" src="https://private-user-images.githubusercontent.com/41707476/373979995-a2dc9e6a-c1cd-4504-821b-f2c35a5f62f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyNDk2ODgsIm5iZiI6MTcyODI0OTM4OCwicGF0aCI6Ii80MTcwNzQ3Ni8zNzM5Nzk5OTUtYTJkYzllNmEtYzFjZC00NTA0LTgyMWItZjJjMzVhNWY2MmYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA2VDIxMTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwMmFiZDc5MWQzNDBhNWUyYmExMGYwZjY4OWMwZDc2NDU3YWU3MjhhYzYyY2Q5NjA1YzI0MTY4ZjM2M2JkZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JCRkNeifPDvoW5iknVFXry0cG_7_j-481lVSMUQhQDk">
> Getting this output now after
> 
> ```
> exo --inference-engine pytorch --run-model llama-3.1-8b
> ```

I'm assuming this comment was meant for #139 ?

AlexCheema commented:
> > <img alt="Screenshot 2024-10-07 at 01 05 01" width="223" src="https://private-user-images.githubusercontent.com/41707476/373979995-a2dc9e6a-c1cd-4504-821b-f2c35a5f62f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyNDk2ODgsIm5iZiI6MTcyODI0OTM4OCwicGF0aCI6Ii80MTcwNzQ3Ni8zNzM5Nzk5OTUtYTJkYzllNmEtYzFjZC00NTA0LTgyMWItZjJjMzVhNWY2MmYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA2VDIxMTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwMmFiZDc5MWQzNDBhNWUyYmExMGYwZjY4OWMwZDc2NDU3YWU3MjhhYzYyY2Q5NjA1YzI0MTY4ZjM2M2JkZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JCRkNeifPDvoW5iknVFXry0cG_7_j-481lVSMUQhQDk">
> > Getting this output now after
> > ```
> > exo --inference-engine pytorch --run-model llama-3.1-8b
> > ```
> 
> I'm assuming this comment was meant for #139 ?

you're right. sorry ignore that :)

AlexCheema commented:
Just ping me when you want me to review the PR again.

snico432 commented:
@AlexCheema I added support for benchmarking on the MLX inference engine. Right now it only benchmarks f32 and f16 calculations because mlx doesn't support matrix multiplication for int8. Not sure how I should proceed with that. Please let me know if this is what you're looking, or if you'd like me to make some changes. I'll move on to Tinygrad afterwards.

AlexCheema commented on exo/topology/mlx_benchmark.py:
why is this in a separate file in `topology`. can we move it into `inference/mlx`?

AlexCheema commented on exo/topology/mlx_benchmark.py:
why is this in a separate file in topology. can we move it into inference/mlx?

snico432 replied:
Sorry, left it there when I was testing. Fixed in my latest commit.

AlexCheema commented on exo/inference/inference_engine.py:
```diff
@@ -15,6 +15,10 @@ async def infer_prompt(self, request_id: str, shard: Shard, prompt: str, image_s
     pass
 
+  @abstractmethod
+  def benchmark_tflops(self):
```
This is going to need to be async otherwise it will block the main thread

snico432 replied:
Fixed!

AlexCheema commented on exo/main.py:
```diff
@@ -74,6 +75,7 @@
   for chatgpt_api_endpoint in chatgpt_api_endpoints:
     print(f" - {terminal_link(chatgpt_api_endpoint)}")
 
+device_capabilities = device_capabilities(inference_engine)
```
I'd prefer a design where `device_capabilities` is asynchronous and called from the discovery module. That way we don't block exo from getting started and the discovery module can fetch the device capabilities when it needs to. Also, if it fails for whatever reason it can be retried. For now lets cache the device capabilities in the discovery module forever but eventually we'd want to "refresh" it by running the benchmark again

snico432 replied:
I think I've now got it cached correctly in the discovery modules (this is my first time working with the async library so I'm not 100% sure)

snico432 commented:
@AlexCheema I created a benchmark for tinygrad, cleaned up the mlx benchmark and attempted to implement your requests from your last review. I also removed the dict from device_capabilites and made all the TFLOPS get calculated dynamically.

AlexCheema reviewed (changes requested):
I merged latest main and made some changes here and there.

I think this is mostly good. Computing DeviceCapabilities is very slow, so I'd prefer if we only compute this once, ideally lazily (i.e. it doesn't slow down startup).

If you can fix that, it should be good to merge.

snico432 commented:
@AlexCheema DeviceCapabilites are now lazily computed. PTAL

AlexCheema commented:
Please fix merge conflicts.

snico432 commented:
@AlexCheema resolved merge conflicts.

snico432 commented:
@AlexCheema PTAL

snico432 commented:
@AlexCheema PTAL
